src/client/method-list-units.c

```diff
@@ -153,18 +157,79 @@ void unit_list_unref(UnitList *unit_list) {
         free(unit_list);
 }
 
-/* Prints 100-characters-wide table */
+void calc_max_col_len(const char *data, unsigned long *max_len) {
+        unsigned long len = strlen(data);
+        if (len > *max_len) {
+                *max_len = len;
+        }
+}
+
+unsigned int calc_line_len(const unsigned int sep_len, const unsigned int num_of_cols, ...) {
+        unsigned int len = 0;
+        va_list args;
+
+        va_start(args, num_of_cols);
+        for (unsigned int i = 0; i < num_of_cols; i++) {
+                len += va_arg(args, unsigned long); // max column length
+                len += sep_len;                     // length of column separator string
+        }
+        va_end(args);
+
+        if (len > 0) {
+                // at least one column exists, remove column separator at the end
+                len -= sep_len;
+        }
+        return len;
+}
+
+void print_header_sep_line(const unsigned int line_len, const char used_char) {
+        for (unsigned int i = 0; i < line_len; i++) {
+                putchar(used_char);
+        }
+        puts("");
+}
+
 void print_unit_list_simple(UnitList *unit_list, const char *glob_filter) {
         UnitInfo *unit = NULL;
-        printf("%-20.20s|%-59.59s|%9s|%9s\n", "NODE", "ID", "ACTIVE", "SUB");
-        printf("====================================================================================================\n");
+        const unsigned int FMT_STR_MAX_LEN = 255;
+        char fmt_str[FMT_STR_MAX_LEN];
+        const char node_title[] = "NODE";
+        const char id_title[] = "ID";
+        const char active_title[] = "ACTIVE";
+        const char sub_title[] = "SUB";
+        const char col_sep[] = " | ";
+
+        unsigned long max_node_len = strlen(node_title);
+        unsigned long max_id_len = strlen(id_title);
+        unsigned long max_active_len = strlen(active_title);
+        unsigned long max_sub_len = strlen(sub_title);
+
+        LIST_FOREACH(units, unit, unit_list->units) {
+                calc_max_col_len(unit->node, &max_node_len);
+                calc_max_col_len(unit->id, &max_id_len);
+                calc_max_col_len(unit->active_state, &max_active_len);
+                calc_max_col_len(unit->sub_state, &max_sub_len);
+        }
+
+        snprintf(fmt_str,
+                 FMT_STR_MAX_LEN,
+                 "%%-%lus%s%%-%lus%s%%-%lus%s%%-%lus\n",
+                 max_node_len,
+                 col_sep,
+                 max_id_len,
+                 col_sep,
+                 max_active_len,
+                 col_sep,
+                 max_sub_len);
+
+        printf(fmt_str, "NODE", "ID", "ACTIVE", "SUB");
+        print_header_sep_line(
```

Comment:
You could use `memset` for this and remove the `print_header_sep_line`:
```C
// +1 to account for the terminating char \0
char sep_line[FMT_STR_MAX_LEN + 1];
// assuming line length is in line_length
memset(&sep_line, '=', sizeof(char) * line_length);
printf("%s", sep_line);
```

Reply:
Done